### DISCUSSION TITLE
Formkit submission in tests causing failed tests

### DISCUSSION THREAD

**Discussion by williamj20** [formkit/formkit#988] [General]
Hi,

I'm writing some tests for some components which are using FormKit. Upon submission of the form in this component, a value in a store gets updated. For example, I expect something like this to work:

```
const form = wrapper.find('form');
await form.trigger('submit');
expect(store.setValue.toHaveBeenCalled());
```
However, this doesn't work (i.e. the test store value is not set after the form is submitted), and I suspect it's because as described [here](https://rickschubert.net/blog/posts/flushing-promises/), nodejs exited before the submission could finish.

So I attempted to remedy this through several methods, such as adding nextTick():

```
const form = wrapper.find('form');
await form.trigger('submit');
await nextTick();
expect(store.setValue.toHaveBeenCalled());
```

or adding flushPromises from the `flush-promises` package:

```
const form = wrapper.find('form');
await form.trigger('submit');
await flushPromises();
expect(store.setValue.toHaveBeenCalled());
```

The solution I stumbled upon is the following, after looking through the code and tests for the formkit package itself:

```
const form = wrapper.find('form');
await form.trigger('submit');
await new Promise((r) => setTimeout(r, 5));
expect(store.setValue.toHaveBeenCalled());
```

My questions are: is this solution correct and expected for me to use? If so, why does this work, but something like flushing promises does not?

**Comment by williamj20** [comment]
After a bit more experimentation, adding two flushPromises calls works, which is very strange to me

```
const form = wrapper.find('form');
await form.trigger('submit');
await flushPromises();
await flushPromises();
expect(store.setValue.toHaveBeenCalled());
```

**Comment by justin-schroeder** [comment]
This actually isn’t terribly complex, but it can be a bit brain melty to read in written text, so i’ll do my best to explain.

All inputs in FormKit are debounced. This means each input has its own lifecycle events. At the most basic level they consist of:

```
Input -> InputHook -> Debounce Timer {20ms by default} -> CommitHook -> Commit
```

This internal debounce timing is fundamental and critical. It allows for inputs to defer expensive computation to a later time and ensures the all the form logic like state tracking is defered to when the main thread has time to work on it while an input is being interacted with. However, there are complexities that arise because of this. For example, what if a a child input is still in performing this lifecycle when a user clicks "submit" — what should happen?

Well the form should wait for those async behaviors to resolve and then proceed with the submit logic. There can be plenty of these behaviors too by the way, like async validation rules — we call these "disturbances" and these are the kinds of frustrating challenges that FormKit attempts to solve "out of the box" — most developers using FormKit wont even know this problem exists and is being solved for them 😂

To handle these edge cases FormKit implements a "disturbance counting" mechanism — meaning each form tree is aware of how many child nodes are currently in a disturbed state. This is done purely with integer logic so it is very fast, it occurs on every keystroke and is one way the form is always aware of its current "state".

When you click the "Submit" button — we check the disturbance count to determine if the form has "settled" or not, and we await all settlement to complete as tabulated by the top of the form tree. During this time there may be multiple async events that occur. `flushPromises()` works by just adding itself to the end of the async call stack, if another promise (or setTimeout resolution) is added to the call stack after `flushPromises()` it will not have run by the time `flushPromises()` is at the top of the call stack.

This means that something as simple as a `setTimeout()` can fall outside the ability of flush promises to catch it. Call it enough times and yes you might get one of your flush calls to be at the true top of the stack.

Unfortunately, as of yet, I don’t know of any great mechanism to work around this — thus the reason you’ll see simple delay calls in our tests pretty frequently too.

Hope that helps at least explain the mechanism behind what is going on here 👍

**Reply by williamj20** [reply]
Thank you!
